### PR TITLE
Updates for 0.12

### PR DESCRIPTION
**Pull request by paulyoung**

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -20,16 +20,17 @@
...
-    "purescript-console": "^3.0.0",
-    "purescript-assert": "^3.0.0",
-    "purescript-quickcheck": "^4.0.0",
-    "purescript-quickcheck-laws": "^3.0.0",
-    "purescript-psci-support": "^3.0.0"
+    "purescript-integers": "^4.0.0",
+    "purescript-console": "^4.1.0",
+    "purescript-assert": "^4.0.0",
+    "purescript-quickcheck": "^5.0.0",
+    "purescript-quickcheck-laws": "paulyoung/purescript-quickcheck-laws#compiler/0.12",
```
Comment (paulyoung) [inline bower.json]: This is blocked by https://github.com/garyb/purescript-quickcheck-laws/pull/39

**Comment by paulyoung** [comment]
I think the current test failure might be legitimate.

```
Binary relations between integers should hold before and after converting to BigInt
100/100 test(s) passed.
100/100 test(s) passed.
52/100 test(s) passed.
/home/travis/build/sharkdp/purescript-bigints/output/Effect.Exception/foreign.js:29
    throw e;
    ^
Error: Test 1 (seed 1030740424) failed: 
Test returned false
    at Object.exports.error (/home/travis/build/sharkdp/purescript-bigints/output/Effect.Exception/foreign.js:8:10)
    at /home/travis/build/sharkdp/purescript-bigints/output/Test.QuickCheck/index.js:139:81
    at /home/travis/build/sharkdp/purescript-bigints/output/Data.Foldable/index.js:95:76
    at /home/travis/build/sharkdp/purescript-bigints/output/Data.Foldable/index.js:292:24
    at /home/travis/build/sharkdp/purescript-bigints/output/Data.Foldable/index.js:417:46
    at /home/travis/build/sharkdp/purescript-bigints/output/Data.Function/index.js:19:24
    at __do (/home/travis/build/sharkdp/purescript-bigints/output/Test.QuickCheck/index.js:138:122)
    at __do (/home/travis/build/sharkdp/purescript-bigints/output/Test.QuickCheck/index.js:169:68)
    at Object.__do [as main] (/home/travis/build/sharkdp/purescript-bigints/output/Test.Main/index.js:139:133)
    at Object.<anonymous> (/tmp/pulp-test118425-4441-1pfwwgw.p8w7.js:1:84)
* ERROR: Subcommand terminated with exit code 1
```

**Comment by paulyoung** [comment]
I can reproduce locally:

```
Binary relations between integers should hold before and after converting to BigInt
100/100 test(s) passed.
100/100 test(s) passed.
54/100 test(s) passed.
Error: Test 3 (seed 556496295) failed:
Test returned false
    at Object.exports.error (/Users/py/projects/sharkdp/purescript-bigints/output/Effect.Exception/foreign.js:8:10)
    at /Users/py/projects/sharkdp/purescript-bigints/output/Test.QuickCheck/index.js:139:81
    at /Users/py/projects/sharkdp/purescript-bigints/output/Data.Foldable/index.js:95:76
    at /Users/py/projects/sharkdp/purescript-bigints/output/Data.Foldable/index.js:292:24
    at /Users/py/projects/sharkdp/purescript-bigints/output/Data.Foldable/index.js:417:46
    at /Users/py/projects/sharkdp/purescript-bigints/output/Data.Function/index.js:19:24
    at __do (/Users/py/projects/sharkdp/purescript-bigints/output/Test.QuickCheck/index.js:138:122)
    at __do (/Users/py/projects/sharkdp/purescript-bigints/output/Test.QuickCheck/index.js:169:68)
    at Object.__do [as main] (/Users/py/projects/sharkdp/purescript-bigints/output/Test.Main/index.js:139:133)
```

**Comment by paulyoung** [comment]
I think these 3 tests are problematic:

```
testBinary (/) (/)
testBinary mod mod
testBinary div div
```

**Comment by sharkdp** [comment]
Thank you very much!

I'll wait for your purescript-quickcheck-laws PR to be merged.

> I think these 3 tests are problematic:
> 
> testBinary (/) (/)
> testBinary mod mod
> testBinary div div

I'm pretty sure this is related to the breaking changes in the Prelude (https://github.com/purescript/purescript-prelude/releases/tag/v4.0.0). In order to still be law-compliant, we will probably have to adapt the `mod` and `div` operations on `BigInt` as well.

**Comment by paulyoung** [comment]
@sharkdp would you might taking over from here to address those test failures? I'm not sure what you'd like to do exactly.

**Comment by sharkdp** [comment]
Sure. I hope my changes make sense (and pass the tests) :smile:

**Review by sharkdp — APPROVED** [review]

**Comment by paulyoung** [comment]
You’re the expert here 😁

**Comment by thomashoneyman** [comment]
@sharkdp @paulyoung Quickcheck laws are in!
https://github.com/garyb/purescript-quickcheck-laws/pull/42

**Comment by sharkdp** [comment]
Thank you very much!

Released as [v4.0.0](https://github.com/sharkdp/purescript-bigints/releases/tag/v4.0.0) (note the breaking changes).

**Comment by hdgarrood** [comment]
Truncating division should give you a lawful EuclideanRing instance. The `Int` instance was only changed because we considered it a more useful default. [My blog post](https://harry.garrood.me/blog/integer-division/) on this topic might be helpful to provide more context.

If `quot` and `rem` aren't passing euclidean ring tests, that may be a bug.

**Comment by hdgarrood** [comment]
> If quot and rem aren't passing euclidean ring tests, that may be a bug.

Ok, I've just looked over the changes you made in https://github.com/sharkdp/purescript-bigints/pull/20/commits/c76ec2f0cc49adb5a4802c179a6fdfb9d904593b and https://github.com/sharkdp/purescript-bigints/pull/20/commits/626dab0f8696f7ac485ef972abf20fc2d9176ae2, and I take this back: I didn't realise that you were testing that e.g. BigInt div matches Int div. (Just thought I'd mention that a truncating EuclideanRing instance _would_ be lawful.) :+1:

**Comment by sharkdp** [comment]
@hdgarrood Thanks - I read your post a while ago :-). I am aware of the different possibilities and we chose to be compliant with `Int`, exactly.

**Pull request merged**